flink-runtime/src/test/java/org/apache/flink/runtime/operators/AbstractOuterJoinTaskTest.java

```diff
@@ -0,0 +1,501 @@
+				}
+			}
+		};
+		taskRunner.start();
+
+		TaskCancelThread tct = new TaskCancelThread(1, taskRunner, this);
+		tct.start();
+
+		try {
+			tct.join();
+			taskRunner.join();
+		} catch(InterruptedException ie) {
+			Assert.fail("Joining threads failed");
+		}
+
+		Assert.assertTrue("Test threw an exception even though it was properly canceled.", success.get());
+	}
+
+	@Test
+	public void testCancelOuterJoinTaskWhileSort2() {
+		int keyCnt = 20;
+		int valCnt = 20;
+
+		setOutput(new DiscardingOutputCollector<Tuple2<Integer, Integer>>());
+		addDriverComparator(this.comparator1);
+		addDriverComparator(this.comparator2);
+		getTaskConfig().setDriverPairComparator(new RuntimePairComparatorFactory());
+		getTaskConfig().setDriverStrategy(driverStrategy);
+		getTaskConfig().setRelativeMemoryDriver(bnljn_frac);
+		setNumFileHandlesForSort(4);
+
+		final AbstractOuterJoinDriver<Tuple2<Integer, Integer>, Tuple2<Integer, Integer>, Tuple2<Integer, Integer>> testTask =
+				getOuterJoinDriver();
+
+		try {
+			addInput(new UniformIntTupleGenerator(keyCnt, valCnt, true), serializer);
+			addInputSorted(new DelayingIterator<>(new InfiniteIntTupleIterator(), 1), this.serializer, this.comparator2.duplicate());
+		} catch (Exception e) {
+			e.printStackTrace();
+			Assert.fail("The test caused an exception.");
+		}
+
+		final AtomicBoolean success = new AtomicBoolean(false);
+
+		Thread taskRunner = new Thread() {
+			@Override
+			public void run() {
+				try {
+					testDriver(testTask, MockJoinStub.class);
+					success.set(true);
+				} catch (Exception ie) {
+					ie.printStackTrace();
+				}
+			}
+		};
+		taskRunner.start();
+
+		TaskCancelThread tct = new TaskCancelThread(1, taskRunner, this);
+		tct.start();
+
+		try {
+			tct.join();
+			taskRunner.join();
+		} catch(InterruptedException ie) {
+			Assert.fail("Joining threads failed");
+		}
+
+		Assert.assertTrue("Test threw an exception even though it was properly canceled.", success.get());
+	}
+
+	@Test
+	public void testCancelOuterJoinTaskWhileRunning() {
+		int keyCnt = 20;
+		int valCnt = 20;
+
+		setOutput(new DiscardingOutputCollector<Tuple2<Integer, Integer>>());
+		addDriverComparator(this.comparator1);
+		addDriverComparator(this.comparator2);
+		getTaskConfig().setDriverPairComparator(new RuntimePairComparatorFactory());
+		getTaskConfig().setDriverStrategy(driverStrategy);
+		getTaskConfig().setRelativeMemoryDriver(bnljn_frac);
+		setNumFileHandlesForSort(4);
+
+		final AbstractOuterJoinDriver<Tuple2<Integer, Integer>, Tuple2<Integer, Integer>, Tuple2<Integer, Integer>> testTask =
+				getOuterJoinDriver();
+
+
+		addInput(new UniformIntTupleGenerator(keyCnt, valCnt, true), serializer);
+		addInput(new UniformIntTupleGenerator(keyCnt, valCnt, true), serializer);
+
+		final AtomicBoolean success = new AtomicBoolean(false);
+
+		Thread taskRunner = new Thread() {
+			@Override
+			public void run() {
+				try {
+					testDriver(testTask, MockDelayingJoinStub.class);
```

Comment:
The reason why it does not cause us trouble on Travis is that the test also succeeds if the testDriver finishes first. But it somehow devaluates the test if it does not test that the cancelling is working properly.